Directly receive remote address when call accept(...)

normanmaurer opened the pull request:
Motivation:

There is a small race in the native transport where an accept(...) may success but a later try to obtain the remote address from the fd may fail is the fd is already closed.

Modifications:

Let accept(...) directly set the remote address.

Result:

No more race possible.

normanmaurer commented:
@trustin @Scottmitch please review

Scottmitch commented on transport-native-epoll/src/main/c/io_netty_channel_epoll_Native.c:
```diff
@@ -1018,21 +1031,27 @@ JNIEXPORT jint JNICALL Java_io_netty_channel_epoll_Native_finishConnect0(JNIEnv*
     jint socketFd;
     int err;
+    struct sockaddr_storage addr;
+    socklen_t address_len;
```
Declare and initialize on the same line?  Good practice (where possible).

normanmaurer replied:
unfortunately this not work with all compilers as far as I know.

Scottmitch replied:
Depends what c standard we are targeting.  However this will not be the first instance to declare/initialize. see [line 299](https://github.com/netty/netty/pull/3440/files#diff-b3c0218865f48def90464f56f26ecce9R299) and other lines too.

normanmaurer replied:
good point... so yeah let me adjust as you said

Scottmitch commented on transport-native-epoll/src/main/java/io/netty/channel/epoll/AcceptedSocket.java:
```diff
@@ -0,0 +1,21 @@
+ */
+package io.netty.channel.epoll;
+
+final class AcceptedSocket {
```
Should we choose a different name for this?  It is not really a "socket" is it?  It doesn't have the FD and can't be used to do sending etc...

normanmaurer replied:
Any good name ?

Scottmitch replied:
`SocketAcceptResults`
`SocketAcceptOutputParameters`
`SocketAcceptMetaData`
Any winners?

normanmaurer replied:
AcceptedAddress ?

Scottmitch replied:
Sure.  I don't think we will be able to stuff anything else in so that is fine.  I'm assuming we can't directly pass the byte[] array through JNI?

normanmaurer replied:
we could just create big-enough byte array that can hold ipv4 / ipv6 / ipv4-mapped addresses but then we still would need a way to know how much of the data in the array we really need to read. I guess we could also just add a marker in the array but I don't think this is any better.

Scottmitch replied:
Yah keep the object and let the one allocating it just size it appropriately.

trustin replied:
Just use the zeroth byte as the length field and document properly?

trustin replied:
+1 for SocketAcceptResult (Results if it accepts many)

normanmaurer replied:
ok let me try to just reuse the array and put the length into the zeroth byte. This way we can safe one allocation per accepted channel

Scottmitch commented on transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollSocketChannel.java:
```diff
@@ -40,12 +40,12 @@
         // See https://github.com/netty/netty/issues/2359
-        remote = Native.remoteAddress(fd);
+        this.remote = remote;
         local = Native.localAddress(fd);
```
Should we add a comment here (or somewhere) that a race condition exists for the local address?  Also indicating that it can't be avoided because the `accept` interface doesn't return the remote address.

normanmaurer replied:
Sure I can add one here... Not sure it helps though ;)

Scottmitch replied:
Just a reminder if/when someone implements another native socket adapter and is referencing this class to avoid this if possible.

Scottmitch commented:
@normanmaurer - Done reviewing.  Just a few comments...good work!

Scottmitch commented:
What about the [remote = ...](https://github.com/netty/netty/blob/accept_remote_address/transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollSocketChannel.java#L187) call on like 187? Can we avoid calling `Native.remoteAddress(fd);`?

normanmaurer commented:
@Scottmitch yeah I think we could just use the remoteAddress that was passed into the method itself. Agree ?

Scottmitch commented:
@normanmaurer - Yes.  We are making sure it is resolved so I think that should be fine.

normanmaurer commented:
Addressed comments and cherry-picked into 4.0 (c21067aab685dbd5909c02c79248fc8a18afbfbd) , 4.1 (88dae15bc20e51f33d22f28ad84b8e62bf7d116d) and master (20e32f62ecc90079d1fe940cc70683617a4b3591)
